### PR TITLE
Add shard lifetime context

### PR DESCRIPTION
**Pull request by dnr**
**What changed?**

Add (Go) context that lives for lifetime of shard context

**Why?**

Useful to track shard lifetime

**How did you test it?**

**Potential risks**

**Is hotfix candidate?**

### CODE REVIEW

File: common/persistence/dataInterfaces.go [hunk 1]
```diff
@@ -197,6 +198,7 @@ type (
 		ShardID          int32
 		CreateIfMissing  bool
```
Comment (wxing1292) [inline common/persistence/dataInterfaces.go:199]: why there is a `create if missing?`

File: common/persistence/dataInterfaces.go [hunk 2]
```diff
@@ -197,6 +198,7 @@ type (
...
 		InitialShardInfo *persistencespb.ShardInfo
+		LifetimeContext  context.Context // cancelled when shard is unloaded
```
Comment (wxing1292) [inline common/persistence/dataInterfaces.go:201]: maybe `LifetimeContext` -> `LifecycleContext`?

File: common/persistence/persistenceInterface.go [hunk 3]
```diff
@@ -215,6 +216,7 @@ type (
 	InternalGetOrCreateShardRequest struct {
 		ShardID         int32
 		CreateShardInfo func() (rangeID int64, shardInfo *commonpb.DataBlob, err error)
```
Comment (wxing1292) [inline common/persistence/persistenceInterface.go:218]: how about `CreateShardInfo func() (rangeID int64, shardInfo *commonpb.DataBlob, err error)` -> default shard info if persistence decides to create?

e.g. 
caller: 
GetOrCreate(shardID int64, defaultShardInfo *commonpb.DataBlob) (*commonpb.DataBlob, error)

File: common/persistence/dataInterfaces.go (hunk 1 above)
Comment (dnr) [reply common/persistence/dataInterfaces.go]: just to avoid changing semantics. removed.

File: common/persistence/dataInterfaces.go (hunk 2 above)
Comment (dnr) [reply common/persistence/dataInterfaces.go]: sure

File: common/persistence/persistenceInterface.go (hunk 3 above)
Comment (dnr) [reply common/persistence/persistenceInterface.go]: keeping this for now

**Review by wxing1292 — APPROVED** [review]

**Pull request merged**